p2p/router.go

```diff
@@ -0,0 +1,557 @@
+			continue
+		}
+
+		msg := proto.Clone(messageType)
+		if err := proto.Unmarshal(bz, msg); err != nil {
+			r.logger.Error("message decoding failed, dropping message", "peer", peerID, "err", err)
+			continue
+		}
+		if wrapper, ok := msg.(Wrapper); ok {
+			msg, err = wrapper.Unwrap()
+			if err != nil {
+				r.logger.Error("failed to unwrap message", "err", err)
+				continue
+			}
+		}
+
+		select {
+		// FIXME: ReceiveMessage() should return ChannelID.
+		case queue.enqueue() <- Envelope{channel: ChannelID(chID), From: peerID, Message: msg}:
+			r.logger.Debug("received message", "peer", peerID, "message", msg)
+		case <-queue.closed():
+			r.logger.Error("channel closed, dropping message", "peer", peerID, "channel", chID)
+		case <-r.stopCh:
+			return nil
+		}
+	}
+}
+
+// sendPeer sends queued messages to a peer.
+func (r *Router) sendPeer(peerID NodeID, conn Connection, queue queue) error {
+	for {
+		select {
+		case envelope := <-queue.dequeue():
+			bz, err := proto.Marshal(envelope.Message)
+			if err != nil {
+				r.logger.Error("failed to marshal message", "peer", peerID, "err", err)
+				continue
+			}
+
+			// FIXME: SendMessage() should take ChannelID.
+			_, err = conn.SendMessage(byte(envelope.channel), bz)
+			if err != nil {
+				return err
+			}
+			r.logger.Debug("sent message", "peer", envelope.To, "message", envelope.Message)
+
+		case <-queue.closed():
+			return nil
+
+		case <-r.stopCh:
+			return nil
+		}
+	}
+}
+
+// SubscribePeerUpdates creates a new peer updates subscription. The caller must
+// consume the peer updates in a timely fashion, since delivery is guaranteed and
+// will block peer connection/disconnection otherwise.
+func (r *Router) SubscribePeerUpdates() (*PeerUpdatesCh, error) {
+	peerUpdates := NewPeerUpdates()
+	r.peerUpdatesMtx.Lock()
+	r.peerUpdatesSubs[peerUpdates] = peerUpdates
+	r.peerUpdatesMtx.Unlock()
+
+	go func() {
+		select {
+		case <-peerUpdates.Done():
+			r.peerUpdatesMtx.Lock()
+			delete(r.peerUpdatesSubs, peerUpdates)
+			r.peerUpdatesMtx.Unlock()
+		case <-r.stopCh:
+		}
+	}()
+	return peerUpdates, nil
+}
+
+// broadcastPeerUpdates broadcasts peer updates received from the router
+// to all subscriptions.
+func (r *Router) broadcastPeerUpdates() {
+	for {
+		select {
+		case peerUpdate := <-r.peerUpdatesCh:
+			subs := []*PeerUpdatesCh{}
+			r.peerUpdatesMtx.RLock()
+			for _, sub := range r.peerUpdatesSubs {
+				subs = append(subs, sub)
+			}
+			r.peerUpdatesMtx.RUnlock()
+
+			for _, sub := range subs {
+				select {
+				case sub.updatesCh <- peerUpdate:
+				case <-sub.doneCh:
+				case <-r.stopCh:
+					return
+				}
+			}
+
+		case <-r.stopCh:
+			return
+		}
+	}
+}
+
+// OnStart implements service.Service.
+func (r *Router) OnStart() error {
+	go r.broadcastPeerUpdates()
+	go r.dialPeers()
+	for _, transport := range r.transports {
+		go r.acceptPeers(transport)
+	}
+	return nil
+}
+
+// OnStop implements service.Service.
+func (r *Router) OnStop() {
```

Comment:
Should we also ensure all `Channel` are closed as well?

Reply:
It's the responsibility of the caller to close channels. I think we should just stop routing, since that's what the caller has requested. The only other thing we can do is panic, I think, since blocking would likely just block forever.

Reply:
Ok, yeah that makes sense.